{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Survival Model for Non-Small Cell Lung Cancer\n",
    "***\n",
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lung cancer is the leading cause of cancer-related deaths worldwide, with an estimated 154,050 deaths in the US alone in 2018. I want to build a predictive model of one-year survival after diagnosis with NSCLC (non-small cell lung cancer) using both clinical and genomic data. Prognostic accuracy for life expectancy is highly valued by patients, their families, and healthcare professionals. It helps determine course of treatment and significantly aids end-of-life decision making.\n",
    "<br/><br/>\n",
    "The simulated dataset was provided by the US Department of Veteran Affairs. I will split the dataset into train, validation, and test sets. Cox proportional hazards regression with elastic net regularization will be used for my survival model. For model evaluation, I have chosen to use both concordance index and average partial log-likelihood."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Description\n",
    "\n",
    "The file clinical.csv contains clinical data on each patient. Its columns are as follows:\n",
    "1.\tID: Unique identifier for the patient.\n",
    "2.\tOutcome: Whether the patient is alive or dead at the follow-up time.\n",
    "3.\tSurvival.Months: The follow-up time in months.\n",
    "4.\tAge: The patient’s age in years at diagnosis.\n",
    "5.\tPrimary.Site: Location of primary tumor.\n",
    "6.\tHistology: Tumor histology.\n",
    "7.\tStage: Stage at diagnosis.\n",
    "8.\tGrade: Tumor grade.\n",
    "9.\tNum.Primaries: Number of primary tumors.\n",
    "10.\tTumor.Size: Size of the tumor at diagnosis.\n",
    "11.\tT: Tumor Stage.\n",
    "12.\tN: Number of metastases to lymph nodes.\n",
    "13.\tM: Number of distant metastases.\n",
    "14.\tRadiation: Whether radiation took place (5) or not (0).\n",
    "15.\tNum.Mutations: The total number of mutations found in the tumor.\n",
    "16.\tNum.Mutated.Genes: The total number of genes with mutation.\n",
    "\n",
    "The file genomics.csv contains information as to which genes were found to have a mutation in each patient’s tumor sequencing data. Only genes with a mutation are listed.\n",
    "1.\tID: Unique identifier for the patient.\n",
    "2.\tGene: The name of the gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Wrangling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All necessary packages are imported, and display options and plotting styles are set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "from lifelines import CoxPHFitter\n",
    "from lifelines.utils.sklearn_adapter import sklearn_adapter\n",
    "from lifelines.calibration import survival_probability_calibration\n",
    "from fancyimpute import IterativeImputer\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from statsmodels.tools.tools import add_constant\n",
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
    "from scipy.stats import percentileofscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 200)\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.width', 100)\n",
    "pd.options.display.multi_sparse = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(style='whitegrid', palette=\"deep\", font_scale=1.1, rc={\"figure.figsize\": [8, 5]})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>The clinical data is imported as a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_c = pd.read_csv('clinical.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Survival.Months</th>\n",
       "      <th>Age</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Num.Primaries</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>M</th>\n",
       "      <th>Radiation</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Primary.Site</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Tumor.Size</th>\n",
       "      <th>Num.Mutated.Genes</th>\n",
       "      <th>Num.Mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alive</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>UNK</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>IV</td>\n",
       "      <td>Left Lower Lobe</td>\n",
       "      <td>Squamous cell carcinoma</td>\n",
       "      <td>1.4</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Dead</td>\n",
       "      <td>19.0</td>\n",
       "      <td>73</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>UNK</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "      <td>IV</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Dead</td>\n",
       "      <td>13.0</td>\n",
       "      <td>72</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Dead</td>\n",
       "      <td>15.0</td>\n",
       "      <td>69</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>1a</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>IA</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Dead</td>\n",
       "      <td>10.0</td>\n",
       "      <td>76</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>UNK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>IIIA</td>\n",
       "      <td>Left Hilar</td>\n",
       "      <td>Large-cell carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Outcome  Survival.Months  Age  Grade  Num.Primaries    T    N    M  Radiation Stage  \\\n",
       "ID                                                                                       \n",
       "1    Alive              9.0   67      4              0  UNK  2.0  NaN          0    IV   \n",
       "2     Dead             19.0   73      2              0  UNK  2.0  0.0          5    IV   \n",
       "3     Dead             13.0   72      3              0    2  2.0  0.0          0  IIIA   \n",
       "4     Dead             15.0   69      9              1   1a  0.0  1.0          0    IA   \n",
       "5     Dead             10.0   76      9              0  UNK  NaN  NaN          0  IIIA   \n",
       "\n",
       "        Primary.Site                Histology  Tumor.Size  Num.Mutated.Genes  Num.Mutations  \n",
       "ID                                                                                           \n",
       "1    Left Lower Lobe  Squamous cell carcinoma         1.4                  8              8  \n",
       "2   Right Upper Lobe           Adenocarcinoma         NaN                  2              2  \n",
       "3   Right Upper Lobe           Adenocarcinoma         1.5                  1              1  \n",
       "4   Right Upper Lobe           Adenocarcinoma         NaN                  4              4  \n",
       "5         Left Hilar     Large-cell carcinoma         NaN                  3              3  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_c.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The clinical DataFrame contains 190 rows (patients) and 15 columns (clinical characteristics). The target variables are Outcome (Survival status at follow-up) and Survival.Months (Follow-up time in months), while the remaining 13 columns are features for prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 190 entries, 1 to 190\n",
      "Data columns (total 15 columns):\n",
      " #   Column             Non-Null Count  Dtype  \n",
      "---  ------             --------------  -----  \n",
      " 0   Outcome            190 non-null    object \n",
      " 1   Survival.Months    190 non-null    float64\n",
      " 2   Age                190 non-null    int64  \n",
      " 3   Grade              190 non-null    int64  \n",
      " 4   Num.Primaries      190 non-null    int64  \n",
      " 5   T                  190 non-null    object \n",
      " 6   N                  125 non-null    float64\n",
      " 7   M                  94 non-null     float64\n",
      " 8   Radiation          190 non-null    int64  \n",
      " 9   Stage              190 non-null    object \n",
      " 10  Primary.Site       190 non-null    object \n",
      " 11  Histology          190 non-null    object \n",
      " 12  Tumor.Size         98 non-null     float64\n",
      " 13  Num.Mutated.Genes  190 non-null    int64  \n",
      " 14  Num.Mutations      190 non-null    int64  \n",
      "dtypes: float64(4), int64(6), object(5)\n",
      "memory usage: 23.8+ KB\n"
     ]
    }
   ],
   "source": [
    "df_c.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Several features contain missing values. N (# of metastases to lymph nodes), M (# of distant metastases) and Tumor.Size (Tumor size at diagnosis) all contain np.NaN values. Meanwhile, T (Tumor stage) contains 'UNK' values, and Grade (Tumor grade) contains 9's.<br/><br/>\n",
    "There are 96 missing values in Grade, 62 in T, 65 in N, 96 in M, and 92 in Tumor.Size.<br/><br/>\n",
    "Some features contain errors. Stage (Stage at diagnosis) contains the value '1B' which is meant to be 'IB', while Primary.Site (Location of primary tumor) contains the value 'Righ Upper Lobe' which is meant to be 'Right Upper Lobe'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Outcome: ['Alive', 'Dead']\n",
      "Survival.Months: [9.0, 9.5, 10.0, 11.0, 13.0, 15.0, 16.0, 18.0, 19.0, 22.0, 23.0, 24.0, 26.0, 29.0, 32.0, 33.0, 34.0, 35.0, 36.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 46.0, 50.0, 71.0]\n",
      "Age: [56, 59, 60, 62, 63, 67, 68, 69, 70, 71, 72, 73, 74, 76, 77, 78, 80, 82, 83, 84]\n",
      "Grade: [2, 3, 4, 9]\n",
      "Num.Primaries: [0, 1]\n",
      "T: ['1', '1a', '1b', '2', '2a', '2b', '3', '4', 'UNK']\n",
      "N: [0.0, 2.0, nan, 1.0, 3.0]\n",
      "M: [nan, 0.0, 1.0]\n",
      "Radiation: [0, 5]\n",
      "Stage: ['1B', 'IA', 'IB', 'IIA', 'IIB', 'IIIA', 'IIIB', 'IV', 'IVB']\n",
      "Primary.Site: ['Both Lung', 'Left Hilar', 'Left Lower Lobe', 'Left Upper Lobe', 'Righ Upper Lobe', 'Right Hilar', 'Right Lower Lobe', 'Right Middle Lobe', 'Right Upper Lobe']\n",
      "Histology: ['Adenocarcinoma', 'Large-cell carcinoma', 'Squamous cell carcinoma']\n",
      "Tumor.Size: [1.4, nan, 1.0, 1.5, 1.6, 1.8, 1.9, 2.0, 2.5, 3.5, 3.6, 4.0, 4.4, 5.3, 5.4, 5.5, 8.0, 8.5, 9.0, 10.0]\n",
      "Num.Mutated.Genes: [0, 1, 2, 3, 4, 5, 6, 7, 8]\n",
      "Num.Mutations: [0, 1, 2, 3, 4, 5, 6, 7, 8]\n"
     ]
    }
   ],
   "source": [
    "# Unique values of each column in df_c\n",
    "for col in df_c:\n",
    "    print(col + ':', sorted(df_c[col].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dead     150\n",
      "Alive     40\n",
      "Name: Outcome, dtype: int64\n",
      "11.0    27\n",
      "10.0    23\n",
      "13.0    21\n",
      "36.0    18\n",
      "32.0    11\n",
      "38.0     9\n",
      "16.0     8\n",
      "33.0     8\n",
      "15.0     7\n",
      "22.0     7\n",
      "9.0      7\n",
      "19.0     6\n",
      "35.0     6\n",
      "23.0     6\n",
      "34.0     4\n",
      "29.0     3\n",
      "9.5      3\n",
      "42.0     3\n",
      "18.0     2\n",
      "39.0     2\n",
      "71.0     2\n",
      "46.0     1\n",
      "50.0     1\n",
      "37.0     1\n",
      "41.0     1\n",
      "40.0     1\n",
      "26.0     1\n",
      "24.0     1\n",
      "Name: Survival.Months, dtype: int64\n",
      "62    27\n",
      "76    26\n",
      "67    22\n",
      "72    20\n",
      "71    14\n",
      "73    13\n",
      "69    13\n",
      "70    10\n",
      "63     8\n",
      "82     7\n",
      "74     7\n",
      "56     5\n",
      "83     4\n",
      "68     4\n",
      "77     4\n",
      "80     2\n",
      "78     1\n",
      "60     1\n",
      "59     1\n",
      "84     1\n",
      "Name: Age, dtype: int64\n",
      "9    96\n",
      "4    43\n",
      "2    29\n",
      "3    22\n",
      "Name: Grade, dtype: int64\n",
      "0    147\n",
      "1     43\n",
      "Name: Num.Primaries, dtype: int64\n",
      "UNK    62\n",
      "3      38\n",
      "1a     26\n",
      "4      23\n",
      "2a     16\n",
      "2      12\n",
      "2b     10\n",
      "1b      2\n",
      "1       1\n",
      "Name: T, dtype: int64\n",
      "2.0    58\n",
      "0.0    52\n",
      "1.0     9\n",
      "3.0     6\n",
      "Name: N, dtype: int64\n",
      "0.0    86\n",
      "1.0     8\n",
      "Name: M, dtype: int64\n",
      "0    127\n",
      "5     63\n",
      "Name: Radiation, dtype: int64\n",
      "IV      45\n",
      "IIIA    43\n",
      "IA      32\n",
      "IVB     25\n",
      "IIIB    24\n",
      "IIB     11\n",
      "IIA      8\n",
      "1B       1\n",
      "IB       1\n",
      "Name: Stage, dtype: int64\n",
      "Right Upper Lobe     53\n",
      "Right Hilar          33\n",
      "Left Hilar           31\n",
      "Right Lower Lobe     25\n",
      "Left Upper Lobe      21\n",
      "Left Lower Lobe      17\n",
      "Both Lung             5\n",
      "Right Middle Lobe     3\n",
      "Righ Upper Lobe       2\n",
      "Name: Primary.Site, dtype: int64\n",
      "Adenocarcinoma             86\n",
      "Squamous cell carcinoma    77\n",
      "Large-cell carcinoma       27\n",
      "Name: Histology, dtype: int64\n",
      "2.0     20\n",
      "1.5     13\n",
      "9.0     10\n",
      "3.6      9\n",
      "10.0     8\n",
      "4.0      7\n",
      "5.5      6\n",
      "8.5      6\n",
      "3.5      2\n",
      "8.0      2\n",
      "1.8      2\n",
      "1.4      2\n",
      "1.9      2\n",
      "5.4      2\n",
      "5.3      2\n",
      "2.5      2\n",
      "1.0      1\n",
      "4.4      1\n",
      "1.6      1\n",
      "Name: Tumor.Size, dtype: int64\n",
      "3    55\n",
      "2    53\n",
      "1    33\n",
      "4    21\n",
      "5    15\n",
      "0     6\n",
      "6     4\n",
      "8     2\n",
      "7     1\n",
      "Name: Num.Mutated.Genes, dtype: int64\n",
      "3    50\n",
      "2    45\n",
      "1    26\n",
      "5    25\n",
      "4    21\n",
      "6    11\n",
      "0     6\n",
      "8     3\n",
      "7     3\n",
      "Name: Num.Mutations, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Values counts for each column in df_c\n",
    "for col in df_c:\n",
    "    print(df_c[col].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All missing values are changed to np.NaN, and the erroneous values are corrected. Also, for the binary variables Outcome and Radiation (whether the patient recieved radiation), I changed the values to [0, 1]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary to find and replace values\n",
    "dic_encode = {'Grade': {9: np.nan},\n",
    "             'Radiation': {5: 1},\n",
    "             'Stage': {'1B': 'IB'},\n",
    "             'T': {'UNK': np.nan},\n",
    "             'Primary.Site': {'Righ Upper Lobe': 'Right Upper Lobe'},\n",
    "             'Outcome': {'Alive': 0, 'Dead': 1}\n",
    "            }\n",
    "\n",
    "df_c.replace(dic_encode, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cancer stages range from 1 to 4, and each stage contains sub-stages of A and B. For the variables T and Stage, several values do not include sub-stages. For consistency purposes, I decided to remove the sub-stage information. I also changed the values of Stage to numerical values, because it is an ordinal variable and will allow me to test its correlation initially."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary to find and replace values\n",
    "dic_encode_2 = {'T': {'1': 1,'1a': 1,'1b': 1,'2': 2,'2a': 2,'2b': 2, '3': 3,'4': 4},\n",
    "              'Stage': {'IA': 1,'IB': 1,'IIA': 2,'IIB': 2,'IIIA': 3,'IIIB': 3, 'IV': 4, 'IVB': 4}\n",
    "             }\n",
    "\n",
    "df_c.replace(dic_encode_2, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Survival.Months</th>\n",
       "      <th>Age</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Num.Primaries</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>M</th>\n",
       "      <th>Radiation</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Tumor.Size</th>\n",
       "      <th>Num.Mutated.Genes</th>\n",
       "      <th>Num.Mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>94.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>128.000000</td>\n",
       "      <td>125.000000</td>\n",
       "      <td>94.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>98.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.789474</td>\n",
       "      <td>22.186842</td>\n",
       "      <td>70.173684</td>\n",
       "      <td>3.148936</td>\n",
       "      <td>0.226316</td>\n",
       "      <td>2.429688</td>\n",
       "      <td>1.144000</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>0.331579</td>\n",
       "      <td>2.910526</td>\n",
       "      <td>4.494898</td>\n",
       "      <td>2.684211</td>\n",
       "      <td>3.084211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.408760</td>\n",
       "      <td>12.420140</td>\n",
       "      <td>6.146909</td>\n",
       "      <td>0.867048</td>\n",
       "      <td>0.419551</td>\n",
       "      <td>1.032416</td>\n",
       "      <td>1.029438</td>\n",
       "      <td>0.280536</td>\n",
       "      <td>0.472024</td>\n",
       "      <td>1.087395</td>\n",
       "      <td>3.050988</td>\n",
       "      <td>1.460327</td>\n",
       "      <td>1.697575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>56.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>71.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.600000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>71.000000</td>\n",
       "      <td>84.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>8.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Outcome  Survival.Months         Age      Grade  Num.Primaries           T           N  \\\n",
       "count  190.000000       190.000000  190.000000  94.000000     190.000000  128.000000  125.000000   \n",
       "mean     0.789474        22.186842   70.173684   3.148936       0.226316    2.429688    1.144000   \n",
       "std      0.408760        12.420140    6.146909   0.867048       0.419551    1.032416    1.029438   \n",
       "min      0.000000         9.000000   56.000000   2.000000       0.000000    1.000000    0.000000   \n",
       "25%      1.000000        11.000000   67.000000   2.000000       0.000000    2.000000    0.000000   \n",
       "50%      1.000000        16.000000   71.000000   3.000000       0.000000    2.000000    2.000000   \n",
       "75%      1.000000        34.000000   74.000000   4.000000       0.000000    3.000000    2.000000   \n",
       "max      1.000000        71.000000   84.000000   4.000000       1.000000    4.000000    3.000000   \n",
       "\n",
       "               M   Radiation       Stage  Tumor.Size  Num.Mutated.Genes  Num.Mutations  \n",
       "count  94.000000  190.000000  190.000000   98.000000         190.000000     190.000000  \n",
       "mean    0.085106    0.331579    2.910526    4.494898           2.684211       3.084211  \n",
       "std     0.280536    0.472024    1.087395    3.050988           1.460327       1.697575  \n",
       "min     0.000000    0.000000    1.000000    1.000000           0.000000       0.000000  \n",
       "25%     0.000000    0.000000    2.000000    2.000000           2.000000       2.000000  \n",
       "50%     0.000000    0.000000    3.000000    3.600000           3.000000       3.000000  \n",
       "75%     0.000000    1.000000    4.000000    8.000000           3.000000       4.000000  \n",
       "max     1.000000    1.000000    4.000000   10.000000           8.000000       8.000000  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_c.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After looking at both feature frequencies, I have decided to combine the values 'Right Middle Lobe' and 'Both Lung' within Primary.Site into a new value named 'Other'. The value 'Other' will have 8 positive occurences, which will pose less problems when splitting the dataset into train and test sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary to find and replace values\n",
    "dic_encode_3 = {'Primary.Site': {'Right Middle Lobe': 'Other','Both Lung': 'Other'}}\n",
    "\n",
    "df_c.replace(dic_encode_3, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The genomic data is now imported as a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g = pd.read_csv('genomics.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>AKT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ALK_Col1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ALK_Col1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ALK_Col2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene\n",
       "ID           \n",
       "1        AKT1\n",
       "158      AKT1\n",
       "88   ALK_Col1\n",
       "132  ALK_Col1\n",
       "18   ALK_Col2"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The genomic DataFrame has 510 rows and one column, but it needs to be changed to have the same structure as the clinical DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(510, 1)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I now change the genomic DataFrame so that each patient's ID is only listed once on the index, while the names of the genes are the column labels. The values of the restructured DataFrame are 1's and 0's indicating whether or not a patient has a mutation of a given gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g['Value'] = 1\n",
    "df_g = df_g.pivot_table(index='ID', columns='Gene', aggfunc=lambda x: int(x.any()), fill_value=0)\n",
    "df_g.columns = df_g.columns.droplevel()\n",
    "df_g.columns.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AKT1</th>\n",
       "      <th>ALK_Col1</th>\n",
       "      <th>ALK_Col2</th>\n",
       "      <th>APC</th>\n",
       "      <th>ATM_Col1</th>\n",
       "      <th>ATM_Col2</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>CCND2</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>CTNNB1</th>\n",
       "      <th>DNMT3A</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>FBXW7</th>\n",
       "      <th>FGFR1</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>FLT4</th>\n",
       "      <th>FOXL2</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>KRAS_Col1</th>\n",
       "      <th>KRAS_Col2</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MET</th>\n",
       "      <th>MLH_Col2</th>\n",
       "      <th>MSH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>NF_Col1</th>\n",
       "      <th>NF_Col2</th>\n",
       "      <th>NF_Col3</th>\n",
       "      <th>NF_Col5</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>NTRK1</th>\n",
       "      <th>PDGFRB</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PIK3CB</th>\n",
       "      <th>POLD_Col2</th>\n",
       "      <th>PTCH1</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>RB1</th>\n",
       "      <th>SMARCA4</th>\n",
       "      <th>SMARCB1</th>\n",
       "      <th>SMO</th>\n",
       "      <th>STK11</th>\n",
       "      <th>TERT</th>\n",
       "      <th>TP53_Col1</th>\n",
       "      <th>TP53_Col2</th>\n",
       "      <th>TSC2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AKT1  ALK_Col1  ALK_Col2  APC  ATM_Col1  ATM_Col2  BRAF  CCND2  CDKN2A  CTNNB1  DNMT3A  EGFR  \\\n",
       "ID                                                                                                 \n",
       "1      1         0         0    0         0         0     0      1       0       0       0     1   \n",
       "2      0         0         0    0         0         0     0      0       0       0       0     0   \n",
       "3      0         0         0    0         0         0     0      0       0       0       0     0   \n",
       "4      0         0         0    0         0         0     0      0       1       0       0     0   \n",
       "5      0         0         0    0         0         0     0      0       0       0       0     0   \n",
       "\n",
       "    ERBB3  ERBB4  ESR1  FBXW7  FGFR1  FGFR3  FLT4  FOXL2  GNAS  HNF1A  KRAS_Col1  KRAS_Col2  \\\n",
       "ID                                                                                            \n",
       "1       0      0     0      0      0      1     0      0     0      0          1          0   \n",
       "2       0      0     0      0      0      0     0      0     0      0          0          0   \n",
       "3       0      0     0      0      0      0     0      0     0      0          0          0   \n",
       "4       0      0     0      0      0      0     0      0     0      0          0          0   \n",
       "5       0      0     0      0      0      0     0      0     0      0          1          0   \n",
       "\n",
       "    MAP2K2  MET  MLH_Col2  MSH2  MSH6  NF_Col1  NF_Col2  NF_Col3  NF_Col5  NOTCH1  NTRK1  PDGFRB  \\\n",
       "ID                                                                                                 \n",
       "1        0    0         0     0     0        0        0        0        0       0      0       1   \n",
       "2        0    0         0     0     0        0        0        0        0       0      0       0   \n",
       "3        0    0         0     0     0        0        0        0        0       0      0       0   \n",
       "4        0    1         0     0     0        0        0        0        0       0      0       0   \n",
       "5        0    0         0     0     0        0        0        0        0       0      0       0   \n",
       "\n",
       "    PIK3CA  PIK3CB  POLD_Col2  PTCH1  PTEN  RB1  SMARCA4  SMARCB1  SMO  STK11  TERT  TP53_Col1  \\\n",
       "ID                                                                                               \n",
       "1        0       0          0      0     0    0        0        0    0      1     0          0   \n",
       "2        0       0          0      0     0    0        0        0    1      0     0          1   \n",
       "3        0       0          0      0     0    0        0        0    0      1     0          0   \n",
       "4        0       1          0      0     0    0        0        0    0      0     0          1   \n",
       "5        0       0          0      0     0    0        0        0    0      0     0          1   \n",
       "\n",
       "    TP53_Col2  TSC2  \n",
       "ID                   \n",
       "1           0     1  \n",
       "2           0     0  \n",
       "3           0     0  \n",
       "4           0     0  \n",
       "5           0     1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The gemomic DataFrame contains 184 rows (patients) and 50 columns (genes). Not all 190 patients from the clinical dataset are included, because the clinical dataset listed 6 patients as having no gene mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(184, 50)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 0]\n"
     ]
    }
   ],
   "source": [
    "# The unique values in all of df_g\n",
    "print(pd.unique(df_g.values.ravel()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TP53_Col1    117\n",
       "KRAS_Col1     55\n",
       "CDKN2A        45\n",
       "TSC2          31\n",
       "MSH2          30\n",
       "STK11         23\n",
       "APC           19\n",
       "PIK3CB        11\n",
       "NF_Col2       10\n",
       "TERT          10\n",
       "SMARCB1        9\n",
       "MET            9\n",
       "SMO            8\n",
       "FBXW7          8\n",
       "TP53_Col2      8\n",
       "GNAS           7\n",
       "MSH6           7\n",
       "NF_Col3        7\n",
       "PTEN           7\n",
       "NTRK1          7\n",
       "PIK3CA         7\n",
       "EGFR           6\n",
       "NF_Col1        5\n",
       "PDGFRB         5\n",
       "POLD_Col2      5\n",
       "FGFR1          4\n",
       "CTNNB1         4\n",
       "RB1            4\n",
       "PTCH1          4\n",
       "ATM_Col1       4\n",
       "NOTCH1         4\n",
       "DNMT3A         3\n",
       "CCND2          2\n",
       "ALK_Col2       2\n",
       "ALK_Col1       2\n",
       "ERBB4          2\n",
       "AKT1           2\n",
       "FGFR3          2\n",
       "FLT4           2\n",
       "FOXL2          2\n",
       "NF_Col5        2\n",
       "ERBB3          1\n",
       "ESR1           1\n",
       "HNF1A          1\n",
       "BRAF           1\n",
       "ATM_Col2       1\n",
       "KRAS_Col2      1\n",
       "MLH_Col2       1\n",
       "SMARCA4        1\n",
       "MAP2K2         1\n",
       "Name: Frequency, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The frequency of each gene mutation\n",
    "freq_g = df_g.sum().sort_values(ascending=False)\n",
    "freq_g.name = 'Frequency'\n",
    "freq_g"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the clinical and genomic DataFrames are combined horizontally into one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_c.join(df_g, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Survival.Months</th>\n",
       "      <th>Age</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Num.Primaries</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>M</th>\n",
       "      <th>Radiation</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Primary.Site</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Tumor.Size</th>\n",
       "      <th>Num.Mutated.Genes</th>\n",
       "      <th>Num.Mutations</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>ALK_Col1</th>\n",
       "      <th>ALK_Col2</th>\n",
       "      <th>APC</th>\n",
       "      <th>ATM_Col1</th>\n",
       "      <th>ATM_Col2</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>CCND2</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>CTNNB1</th>\n",
       "      <th>DNMT3A</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>FBXW7</th>\n",
       "      <th>FGFR1</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>FLT4</th>\n",
       "      <th>FOXL2</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>KRAS_Col1</th>\n",
       "      <th>KRAS_Col2</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MET</th>\n",
       "      <th>MLH_Col2</th>\n",
       "      <th>MSH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>NF_Col1</th>\n",
       "      <th>NF_Col2</th>\n",
       "      <th>NF_Col3</th>\n",
       "      <th>NF_Col5</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>NTRK1</th>\n",
       "      <th>PDGFRB</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PIK3CB</th>\n",
       "      <th>POLD_Col2</th>\n",
       "      <th>PTCH1</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>RB1</th>\n",
       "      <th>SMARCA4</th>\n",
       "      <th>SMARCB1</th>\n",
       "      <th>SMO</th>\n",
       "      <th>STK11</th>\n",
       "      <th>TERT</th>\n",
       "      <th>TP53_Col1</th>\n",
       "      <th>TP53_Col2</th>\n",
       "      <th>TSC2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>Left Lower Lobe</td>\n",
       "      <td>Squamous cell carcinoma</td>\n",
       "      <td>1.4</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>19.0</td>\n",
       "      <td>73</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>13.0</td>\n",
       "      <td>72</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Right Upper Lobe</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>10.0</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Left Hilar</td>\n",
       "      <td>Large-cell carcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Outcome  Survival.Months  Age  Grade  Num.Primaries    T    N    M  Radiation  Stage  \\\n",
       "ID                                                                                         \n",
       "1         0              9.0   67    4.0              0  NaN  2.0  NaN          0      4   \n",
       "2         1             19.0   73    2.0              0  NaN  2.0  0.0          1      4   \n",
       "3         1             13.0   72    3.0              0  2.0  2.0  0.0          0      3   \n",
       "4         1             15.0   69    NaN              1  1.0  0.0  1.0          0      1   \n",
       "5         1             10.0   76    NaN              0  NaN  NaN  NaN          0      3   \n",
       "\n",
       "        Primary.Site                Histology  Tumor.Size  Num.Mutated.Genes  Num.Mutations  AKT1  \\\n",
       "ID                                                                                                  \n",
       "1    Left Lower Lobe  Squamous cell carcinoma         1.4                  8              8   1.0   \n",
       "2   Right Upper Lobe           Adenocarcinoma         NaN                  2              2   0.0   \n",
       "3   Right Upper Lobe           Adenocarcinoma         1.5                  1              1   0.0   \n",
       "4   Right Upper Lobe           Adenocarcinoma         NaN                  4              4   0.0   \n",
       "5         Left Hilar     Large-cell carcinoma         NaN                  3              3   0.0   \n",
       "\n",
       "    ALK_Col1  ALK_Col2  APC  ATM_Col1  ATM_Col2  BRAF  CCND2  CDKN2A  CTNNB1  DNMT3A  EGFR  ERBB3  \\\n",
       "ID                                                                                                  \n",
       "1        0.0       0.0  0.0       0.0       0.0   0.0    1.0     0.0     0.0     0.0   1.0    0.0   \n",
       "2        0.0       0.0  0.0       0.0       0.0   0.0    0.0     0.0     0.0     0.0   0.0    0.0   \n",
       "3        0.0       0.0  0.0       0.0       0.0   0.0    0.0     0.0     0.0     0.0   0.0    0.0   \n",
       "4        0.0       0.0  0.0       0.0       0.0   0.0    0.0     1.0     0.0     0.0   0.0    0.0   \n",
       "5        0.0       0.0  0.0       0.0       0.0   0.0    0.0     0.0     0.0     0.0   0.0    0.0   \n",
       "\n",
       "    ERBB4  ESR1  FBXW7  FGFR1  FGFR3  FLT4  FOXL2  GNAS  HNF1A  KRAS_Col1  KRAS_Col2  MAP2K2  MET  \\\n",
       "ID                                                                                                  \n",
       "1     0.0   0.0    0.0    0.0    1.0   0.0    0.0   0.0    0.0        1.0        0.0     0.0  0.0   \n",
       "2     0.0   0.0    0.0    0.0    0.0   0.0    0.0   0.0    0.0        0.0        0.0     0.0  0.0   \n",
       "3     0.0   0.0    0.0    0.0    0.0   0.0    0.0   0.0    0.0        0.0        0.0     0.0  0.0   \n",
       "4     0.0   0.0    0.0    0.0    0.0   0.0    0.0   0.0    0.0        0.0        0.0     0.0  1.0   \n",
       "5     0.0   0.0    0.0    0.0    0.0   0.0    0.0   0.0    0.0        1.0        0.0     0.0  0.0   \n",
       "\n",
       "    MLH_Col2  MSH2  MSH6  NF_Col1  NF_Col2  NF_Col3  NF_Col5  NOTCH1  NTRK1  PDGFRB  PIK3CA  \\\n",
       "ID                                                                                            \n",
       "1        0.0   0.0   0.0      0.0      0.0      0.0      0.0     0.0    0.0     1.0     0.0   \n",
       "2        0.0   0.0   0.0      0.0      0.0      0.0      0.0     0.0    0.0     0.0     0.0   \n",
       "3        0.0   0.0   0.0      0.0      0.0      0.0      0.0     0.0    0.0     0.0     0.0   \n",
       "4        0.0   0.0   0.0      0.0      0.0      0.0      0.0     0.0    0.0     0.0     0.0   \n",
       "5        0.0   0.0   0.0      0.0      0.0      0.0      0.0     0.0    0.0     0.0     0.0   \n",
       "\n",
       "    PIK3CB  POLD_Col2  PTCH1  PTEN  RB1  SMARCA4  SMARCB1  SMO  STK11  TERT  TP53_Col1  TP53_Col2  \\\n",
       "ID                                                                                                  \n",
       "1      0.0        0.0    0.0   0.0  0.0      0.0      0.0  0.0    1.0   0.0        0.0        0.0   \n",
       "2      0.0        0.0    0.0   0.0  0.0      0.0      0.0  1.0    0.0   0.0        1.0        0.0   \n",
       "3      0.0        0.0    0.0   0.0  0.0      0.0      0.0  0.0    1.0   0.0        0.0        0.0   \n",
       "4      1.0        0.0    0.0   0.0  0.0      0.0      0.0  0.0    0.0   0.0        1.0        0.0   \n",
       "5      0.0        0.0    0.0   0.0  0.0      0.0      0.0  0.0    0.0   0.0        1.0        0.0   \n",
       "\n",
       "    TSC2  \n",
       "ID        \n",
       "1    1.0  \n",
       "2    0.0  \n",
       "3    0.0  \n",
       "4    0.0  \n",
       "5    1.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The np.NaN values of the 6 patients with no genetic mutations are filled with zeros."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[:,'AKT1':] = df.loc[:,'AKT1':].fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The full dataset now has 190 rows (patients) and 65 columns (clinical and genomic characteristics)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(190, 65)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are missing values in the following columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Grade', 'T', 'N', 'M', 'Tumor.Size']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_cols = df.columns[df.isna().any()].tolist()\n",
    "missing_cols"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before imputing missing values and performing feature selection, the dataset will be split into train, validation and test sets to prevent data leakage. I will use stratification so that the train, validation and test sets are more likely to be representative of the dataset as a whole. I want each subset to have a similar distribution of values from both Outlook and Survival.Months.<br/><br/>\n",
    "I put the values of Survival.Months into 3 separate bins divided by the approximate tertile values for 33 and 67 percentile. Then I combine the former labels of the Outcome values ('Alive' and 'Dead') with the new group labels of the Survival.Months values. This will result in each patient being placed in one of 6 distinct groups (Alive-1, Alive-2, Alive-3, Dead-1, Dead-2, and Dead-3). Next, StratifiedShuffleSplit is used to create random train, validation and test indices stratified based on the distributions of the newly created groups. The train, validation and test sets are later created with these indices. The train-validation-test split percentage is 70-15-15."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create 6 groups for the patients so that they can be stratified by both Outcome and Survival.Months\n",
    "months = df.loc[:, 'Survival.Months']\n",
    "bins = np.array([min(months), months.quantile(1/3), months.quantile(2/3), max(months) + 1])\n",
    "months_binned = np.digitize(months, bins)\n",
    "months_binned = pd.Series(months_binned).astype(str)\n",
    "outcome = df['Outcome'].replace({0: 'Alive', 1: 'Dead'})\n",
    "patients_grouped = outcome + '-' + months_binned.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dead-2     52\n",
       "Dead-1     50\n",
       "Dead-3     48\n",
       "Alive-3    20\n",
       "Alive-2    10\n",
       "Alive-1    10\n",
       "Name: Outcome, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Distribution of groups\n",
    "patients_grouped.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create stratified train, validation and test set indices using StratifiedShuffleSplit twice\n",
    "sss1 = StratifiedShuffleSplit(n_splits=1, test_size=0.3, random_state=88)\n",
    "\n",
    "for train_index, val_test_index in sss1.split(patients_grouped, patients_grouped):\n",
    "    pg_val_test = patients_grouped.iloc[val_test_index]\n",
    "    train_index = train_index + 1    \n",
    "    \n",
    "sss2 = StratifiedShuffleSplit(n_splits=1, test_size=0.5, random_state=88)\n",
    "\n",
    "for val_index, test_index in sss2.split(pg_val_test, pg_val_test):\n",
    "    pg_val = pg_val_test.iloc[val_index]\n",
    "    pg_test = pg_val_test.iloc[test_index]\n",
    "    val_index = pg_val.index\n",
    "    test_index = pg_test.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataset has 3 ordinal features (Grade, T, and Stage) and 2 categorical features (Primary.Site and Histology), and I will create dummy variables for these 5 features before I begin modelling.<br/><br/>\n",
    "However, there are still missing values within several features (Grade, T, N, M, and Tumor.Size). I will use iterative multivariate feature imputation to fill in the missing values. The imputer will model each feature with missing values as a function of the other features, and then use that estimate for imputation. All of the features must have numerical values before I begin imputation. Neither of the categorical features have missing values, nor does the ordinal variable Stage.<br/><br/>\n",
    "So, I will first create dummy variables for these 3 features (Stage, Primary.Site, and Histology). I will drop the first category/unique value of each feature, since only k - 1 indicator variables are needed to represent the k categories of a categorial variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Covert 3 features into dummy variables\n",
    "df = pd.get_dummies(df, columns=['Stage','Primary.Site','Histology'], drop_first=True).copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The train, validation and test sets are created with the indices generated previously."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "train = df.loc[train_index]\n",
    "val = df.loc[val_index]\n",
    "test = df.loc[test_index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will keep the target variables (Outcome and Survival.Months) separate while performing imputation, so that the imputed test data is not influenced by this information. The training set is divided into y_train (containing just Outcome and Survival.Months) and X_train which contains the predictor variables. The validation and test sets are split similarly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = train.drop(['Outcome','Survival.Months'],axis=1)\n",
    "y_train = train.loc[:, ['Outcome','Survival.Months']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_val = val.drop(['Outcome','Survival.Months'],axis=1)\n",
    "y_val = val.loc[:, ['Outcome','Survival.Months']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_test = test.drop(['Outcome','Survival.Months'],axis=1)\n",
    "y_test = test.loc[:, ['Outcome','Survival.Months']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The imputer is initialized with a specified random seed for reproducibility. It is then fit on X_train. The imputer then transforms/imputes X_train and the result is assigned to X_train_imputed. The imputer remains fitted on X_train, and the same process of transformation and assingment takes place for both the validation and test set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputer = IterativeImputer(random_state=88)\n",
    "\n",
    "X_train_imputed = X_train.copy()\n",
    "X_val_imputed = X_val.copy()\n",
    "X_test_imputed = X_test.copy()\n",
    "\n",
    "imputer.fit(X_train)\n",
    "\n",
    "X_train_imputed.iloc[:, :] = imputer.transform(X_train)\n",
    "\n",
    "X_val_imputed.iloc[:, :] = imputer.transform(X_val)\n",
    "\n",
    "X_test_imputed.iloc[:, :] = imputer.transform(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "y_train and X_train are joined to form the imputed training set, train_imputed. The imputed validation and test sets are formed in the same way.<br/><br/>\n",
    "The full imputed data set, df_imputed, is formed by concatenating train_imputed, val_imputed and test_imputed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_imputed = y_train.join(X_train_imputed, how='left')\n",
    "val_imputed = y_val.join(X_val_imputed, how='left')\n",
    "test_imputed = y_test.join(X_test_imputed, how='left')\n",
    "\n",
    "# The rows of the resulting DataFrame are sorted\n",
    "df_imputed = pd.concat([train_imputed, val_imputed, test_imputed]).sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The imputer imputes values with several decimals, so the imputed values need to be rounded. The values of Tumor.Size are rounded to one decimal, while the remaining imputed features are rounded to integers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in ['Grade','T','N','M']:\n",
    "    df_imputed.loc[:, col] = df_imputed[col].round()\n",
    "    \n",
    "df_imputed.loc[:, 'Tumor.Size'] = df_imputed['Tumor.Size'].round(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The value frequencies of the features with missing values are shown both before and after imputation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial:\n",
      "4.0    43\n",
      "2.0    29\n",
      "3.0    22\n",
      "Name: Grade, dtype: int64 \n",
      "\n",
      "Post-imputation:\n",
      "3.0    84\n",
      "4.0    65\n",
      "2.0    41\n",
      "Name: Grade, dtype: int64 \n",
      "\n",
      "Initial:\n",
      "3.0    38\n",
      "2.0    38\n",
      "1.0    29\n",
      "4.0    23\n",
      "Name: T, dtype: int64 \n",
      "\n",
      "Post-imputation:\n",
      "2.0    100\n",
      "3.0     38\n",
      "1.0     29\n",
      "4.0     23\n",
      "Name: T, dtype: int64 \n",
      "\n",
      "Initial:\n",
      "2.0    58\n",
      "0.0    52\n",
      "1.0     9\n",
      "3.0     6\n",
      "Name: N, dtype: int64 \n",
      "\n",
      "Post-imputation:\n",
      "2.0    68\n",
      "1.0    64\n",
      "0.0    52\n",
      "3.0     6\n",
      "Name: N, dtype: int64 \n",
      "\n",
      "Initial:\n",
      "0.0    86\n",
      "1.0     8\n",
      "Name: M, dtype: int64 \n",
      "\n",
      "Post-imputation:\n",
      "0.0    182\n",
      "1.0      8\n",
      "Name: M, dtype: int64 \n",
      "\n",
      "Initial:\n",
      "2.0     20\n",
      "1.5     13\n",
      "9.0     10\n",
      "3.6      9\n",
      "10.0     8\n",
      "4.0      7\n",
      "5.5      6\n",
      "8.5      6\n",
      "3.5      2\n",
      "8.0      2\n",
      "1.8      2\n",
      "1.4      2\n",
      "1.9      2\n",
      "5.4      2\n",
      "5.3      2\n",
      "2.5      2\n",
      "1.0      1\n",
      "4.4      1\n",
      "1.6      1\n",
      "Name: Tumor.Size, dtype: int64 \n",
      "\n",
      "Post-imputation:\n",
      "4.6     92\n",
      "2.0     20\n",
      "1.5     13\n",
      "9.0     10\n",
      "3.6      9\n",
      "10.0     8\n",
      "4.0      7\n",
      "5.5      6\n",
      "8.5      6\n",
      "3.5      2\n",
      "8.0      2\n",
      "1.8      2\n",
      "1.4      2\n",
      "1.9      2\n",
      "5.4      2\n",
      "5.3      2\n",
      "2.5      2\n",
      "1.0      1\n",
      "4.4      1\n",
      "1.6      1\n",
      "Name: Tumor.Size, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for col in missing_cols:\n",
    "    print('Initial:')\n",
    "    print(df[col].value_counts(), '\\n')\n",
    "    print('Post-imputation:')\n",
    "    print(df_imputed[col].value_counts(), '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the value type of 2 columns\n",
    "df_imputed.loc[:, ['Grade','T']] = df_imputed.loc[:, ['Grade','T']].astype(np.int64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two ordinal values orginally contained missing values, and they have been imputed. I will now create dummy variables for these features (Grade and T), dopping the first unique value of each feature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_imputed = pd.get_dummies(df_imputed, columns=['Grade','T'], drop_first=True)\n",
    "\n",
    "# Change all values types to float\n",
    "df_imputed = df_imputed.astype(np.float64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The train, validation and test sets are taken from the current version of the dataset, df_imputed, by using their indices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "train = df_imputed.loc[train.index].copy()\n",
    "val = df_imputed.loc[val.index].copy()\n",
    "test = df_imputed.loc[test.index].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are now 76 variables in the dataset due to the creation of several dummy/indicator variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "76"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Survival.Months</th>\n",
       "      <th>Age</th>\n",
       "      <th>Num.Primaries</th>\n",
       "      <th>N</th>\n",
       "      <th>M</th>\n",
       "      <th>Radiation</th>\n",
       "      <th>Tumor.Size</th>\n",
       "      <th>Num.Mutated.Genes</th>\n",
       "      <th>Num.Mutations</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>ALK_Col1</th>\n",
       "      <th>ALK_Col2</th>\n",
       "      <th>APC</th>\n",
       "      <th>ATM_Col1</th>\n",
       "      <th>ATM_Col2</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>CCND2</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>CTNNB1</th>\n",
       "      <th>DNMT3A</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>FBXW7</th>\n",
       "      <th>FGFR1</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>FLT4</th>\n",
       "      <th>FOXL2</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>KRAS_Col1</th>\n",
       "      <th>KRAS_Col2</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MET</th>\n",
       "      <th>MLH_Col2</th>\n",
       "      <th>MSH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>NF_Col1</th>\n",
       "      <th>NF_Col2</th>\n",
       "      <th>NF_Col3</th>\n",
       "      <th>NF_Col5</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>NTRK1</th>\n",
       "      <th>PDGFRB</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PIK3CB</th>\n",
       "      <th>POLD_Col2</th>\n",
       "      <th>PTCH1</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>RB1</th>\n",
       "      <th>SMARCA4</th>\n",
       "      <th>SMARCB1</th>\n",
       "      <th>SMO</th>\n",
       "      <th>STK11</th>\n",
       "      <th>TERT</th>\n",
       "      <th>TP53_Col1</th>\n",
       "      <th>TP53_Col2</th>\n",
       "      <th>TSC2</th>\n",
       "      <th>Stage_2</th>\n",
       "      <th>Stage_3</th>\n",
       "      <th>Stage_4</th>\n",
       "      <th>Primary.Site_Left Lower Lobe</th>\n",
       "      <th>Primary.Site_Left Upper Lobe</th>\n",
       "      <th>Primary.Site_Other</th>\n",
       "      <th>Primary.Site_Right Hilar</th>\n",
       "      <th>Primary.Site_Right Lower Lobe</th>\n",
       "      <th>Primary.Site_Right Upper Lobe</th>\n",
       "      <th>Histology_Large-cell carcinoma</th>\n",
       "      <th>Histology_Squamous cell carcinoma</th>\n",
       "      <th>Grade_3</th>\n",
       "      <th>Grade_4</th>\n",
       "      <th>T_2</th>\n",
       "      <th>T_3</th>\n",
       "      <th>T_4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Outcome  Survival.Months   Age  Num.Primaries    N    M  Radiation  Tumor.Size  \\\n",
       "ID                                                                                   \n",
       "1       0.0              9.0  67.0            0.0  2.0  0.0        0.0         1.4   \n",
       "2       1.0             19.0  73.0            0.0  2.0  0.0        1.0         4.6   \n",
       "3       1.0             13.0  72.0            0.0  2.0  0.0        0.0         1.5   \n",
       "4       1.0             15.0  69.0            1.0  0.0  1.0        0.0         4.6   \n",
       "5       1.0             10.0  76.0            0.0  1.0  0.0        0.0         4.6   \n",
       "\n",
       "    Num.Mutated.Genes  Num.Mutations  AKT1  ALK_Col1  ALK_Col2  APC  ATM_Col1  ATM_Col2  BRAF  \\\n",
       "ID                                                                                              \n",
       "1                 8.0            8.0   1.0       0.0       0.0  0.0       0.0       0.0   0.0   \n",
       "2                 2.0            2.0   0.0       0.0       0.0  0.0       0.0       0.0   0.0   \n",
       "3                 1.0            1.0   0.0       0.0       0.0  0.0       0.0       0.0   0.0   \n",
       "4                 4.0            4.0   0.0       0.0       0.0  0.0       0.0       0.0   0.0   \n",
       "5                 3.0            3.0   0.0       0.0       0.0  0.0       0.0       0.0   0.0   \n",
       "\n",
       "    CCND2  CDKN2A  CTNNB1  DNMT3A  EGFR  ERBB3  ERBB4  ESR1  FBXW7  FGFR1  FGFR3  FLT4  FOXL2  \\\n",
       "ID                                                                                              \n",
       "1     1.0     0.0     0.0     0.0   1.0    0.0    0.0   0.0    0.0    0.0    1.0   0.0    0.0   \n",
       "2     0.0     0.0     0.0     0.0   0.0    0.0    0.0   0.0    0.0    0.0    0.0   0.0    0.0   \n",
       "3     0.0     0.0     0.0     0.0   0.0    0.0    0.0   0.0    0.0    0.0    0.0   0.0    0.0   \n",
       "4     0.0     1.0     0.0     0.0   0.0    0.0    0.0   0.0    0.0    0.0    0.0   0.0    0.0   \n",
       "5     0.0     0.0     0.0     0.0   0.0    0.0    0.0   0.0    0.0    0.0    0.0   0.0    0.0   \n",
       "\n",
       "    GNAS  HNF1A  KRAS_Col1  KRAS_Col2  MAP2K2  MET  MLH_Col2  MSH2  MSH6  NF_Col1  NF_Col2  \\\n",
       "ID                                                                                           \n",
       "1    0.0    0.0        1.0        0.0     0.0  0.0       0.0   0.0   0.0      0.0      0.0   \n",
       "2    0.0    0.0        0.0        0.0     0.0  0.0       0.0   0.0   0.0      0.0      0.0   \n",
       "3    0.0    0.0        0.0        0.0     0.0  0.0       0.0   0.0   0.0      0.0      0.0   \n",
       "4    0.0    0.0        0.0        0.0     0.0  1.0       0.0   0.0   0.0      0.0      0.0   \n",
       "5    0.0    0.0        1.0        0.0     0.0  0.0       0.0   0.0   0.0      0.0      0.0   \n",
       "\n",
       "    NF_Col3  NF_Col5  NOTCH1  NTRK1  PDGFRB  PIK3CA  PIK3CB  POLD_Col2  PTCH1  PTEN  RB1  SMARCA4  \\\n",
       "ID                                                                                                  \n",
       "1       0.0      0.0     0.0    0.0     1.0     0.0     0.0        0.0    0.0   0.0  0.0      0.0   \n",
       "2       0.0      0.0     0.0    0.0     0.0     0.0     0.0        0.0    0.0   0.0  0.0      0.0   \n",
       "3       0.0      0.0     0.0    0.0     0.0     0.0     0.0        0.0    0.0   0.0  0.0      0.0   \n",
       "4       0.0      0.0     0.0    0.0     0.0     0.0     1.0        0.0    0.0   0.0  0.0      0.0   \n",
       "5       0.0      0.0     0.0    0.0     0.0     0.0     0.0        0.0    0.0   0.0  0.0      0.0   \n",
       "\n",
       "    SMARCB1  SMO  STK11  TERT  TP53_Col1  TP53_Col2  TSC2  Stage_2  Stage_3  Stage_4  \\\n",
       "ID                                                                                     \n",
       "1       0.0  0.0    1.0   0.0        0.0        0.0   1.0      0.0      0.0      1.0   \n",
       "2       0.0  1.0    0.0   0.0        1.0        0.0   0.0      0.0      0.0      1.0   \n",
       "3       0.0  0.0    1.0   0.0        0.0        0.0   0.0      0.0      1.0      0.0   \n",
       "4       0.0  0.0    0.0   0.0        1.0        0.0   0.0      0.0      0.0      0.0   \n",
       "5       0.0  0.0    0.0   0.0        1.0        0.0   1.0      0.0      1.0      0.0   \n",
       "\n",
       "    Primary.Site_Left Lower Lobe  Primary.Site_Left Upper Lobe  Primary.Site_Other  \\\n",
       "ID                                                                                   \n",
       "1                            1.0                           0.0                 0.0   \n",
       "2                            0.0                           0.0                 0.0   \n",
       "3                            0.0                           0.0                 0.0   \n",
       "4                            0.0                           0.0                 0.0   \n",
       "5                            0.0                           0.0                 0.0   \n",
       "\n",
       "    Primary.Site_Right Hilar  Primary.Site_Right Lower Lobe  Primary.Site_Right Upper Lobe  \\\n",
       "ID                                                                                           \n",
       "1                        0.0                            0.0                            0.0   \n",
       "2                        0.0                            0.0                            1.0   \n",
       "3                        0.0                            0.0                            1.0   \n",
       "4                        0.0                            0.0                            1.0   \n",
       "5                        0.0                            0.0                            0.0   \n",
       "\n",
       "    Histology_Large-cell carcinoma  Histology_Squamous cell carcinoma  Grade_3  Grade_4  T_2  T_3  \\\n",
       "ID                                                                                                  \n",
       "1                              0.0                                1.0      0.0      1.0  1.0  0.0   \n",
       "2                              0.0                                0.0      0.0      0.0  1.0  0.0   \n",
       "3                              0.0                                0.0      1.0      0.0  1.0  0.0   \n",
       "4                              0.0                                0.0      1.0      0.0  0.0  0.0   \n",
       "5                              1.0                                0.0      1.0      0.0  1.0  0.0   \n",
       "\n",
       "    T_4  \n",
       "ID       \n",
       "1   0.0  \n",
       "2   0.0  \n",
       "3   0.0  \n",
       "4   0.0  \n",
       "5   0.0  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_imputed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Survival.Months</th>\n",
       "      <th>Age</th>\n",
       "      <th>Num.Primaries</th>\n",
       "      <th>N</th>\n",
       "      <th>M</th>\n",
       "      <th>Radiation</th>\n",
       "      <th>Tumor.Size</th>\n",
       "      <th>Num.Mutated.Genes</th>\n",
       "      <th>Num.Mutations</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>ALK_Col1</th>\n",
       "      <th>ALK_Col2</th>\n",
       "      <th>APC</th>\n",
       "      <th>ATM_Col1</th>\n",
       "      <th>ATM_Col2</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>CCND2</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>CTNNB1</th>\n",
       "      <th>DNMT3A</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>ESR1</th>\n",
       "      <th>FBXW7</th>\n",
       "      <th>FGFR1</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>FLT4</th>\n",
       "      <th>FOXL2</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>KRAS_Col1</th>\n",
       "      <th>KRAS_Col2</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MET</th>\n",
       "      <th>MLH_Col2</th>\n",
       "      <th>MSH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>NF_Col1</th>\n",
       "      <th>NF_Col2</th>\n",
       "      <th>NF_Col3</th>\n",
       "      <th>NF_Col5</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>NTRK1</th>\n",
       "      <th>PDGFRB</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PIK3CB</th>\n",
       "      <th>POLD_Col2</th>\n",
       "      <th>PTCH1</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>RB1</th>\n",
       "      <th>SMARCA4</th>\n",
       "      <th>SMARCB1</th>\n",
       "      <th>SMO</th>\n",
       "      <th>STK11</th>\n",
       "      <th>TERT</th>\n",
       "      <th>TP53_Col1</th>\n",
       "      <th>TP53_Col2</th>\n",
       "      <th>TSC2</th>\n",
       "      <th>Stage_2</th>\n",
       "      <th>Stage_3</th>\n",
       "      <th>Stage_4</th>\n",
       "      <th>Primary.Site_Left Lower Lobe</th>\n",
       "      <th>Primary.Site_Left Upper Lobe</th>\n",
       "      <th>Primary.Site_Other</th>\n",
       "      <th>Primary.Site_Right Hilar</th>\n",
       "      <th>Primary.Site_Right Lower Lobe</th>\n",
       "      <th>Primary.Site_Right Upper Lobe</th>\n",
       "      <th>Histology_Large-cell carcinoma</th>\n",
       "      <th>Histology_Squamous cell carcinoma</th>\n",
       "      <th>Grade_3</th>\n",
       "      <th>Grade_4</th>\n",
       "      <th>T_2</th>\n",
       "      <th>T_3</th>\n",
       "      <th>T_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>190.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.789474</td>\n",
       "      <td>22.186842</td>\n",
       "      <td>70.173684</td>\n",
       "      <td>0.226316</td>\n",
       "      <td>1.147368</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.331579</td>\n",
       "      <td>4.545789</td>\n",
       "      <td>2.684211</td>\n",
       "      <td>3.084211</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.236842</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.015789</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.289474</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.047368</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.010526</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.057895</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.036842</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.047368</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.121053</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.615789</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.163158</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.352632</td>\n",
       "      <td>0.368421</td>\n",
       "      <td>0.089474</td>\n",
       "      <td>0.110526</td>\n",
       "      <td>0.042105</td>\n",
       "      <td>0.173684</td>\n",
       "      <td>0.131579</td>\n",
       "      <td>0.289474</td>\n",
       "      <td>0.142105</td>\n",
       "      <td>0.405263</td>\n",
       "      <td>0.442105</td>\n",
       "      <td>0.342105</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.121053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.408760</td>\n",
       "      <td>12.420140</td>\n",
       "      <td>6.146909</td>\n",
       "      <td>0.419551</td>\n",
       "      <td>0.860275</td>\n",
       "      <td>0.201360</td>\n",
       "      <td>0.472024</td>\n",
       "      <td>2.186359</td>\n",
       "      <td>1.460327</td>\n",
       "      <td>1.697575</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.300793</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.426268</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.124990</td>\n",
       "      <td>0.175338</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.201360</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.454716</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.212987</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.365606</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.160496</td>\n",
       "      <td>0.223887</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.102326</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.160496</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.234161</td>\n",
       "      <td>0.160496</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.188872</td>\n",
       "      <td>0.143939</td>\n",
       "      <td>0.072548</td>\n",
       "      <td>0.212987</td>\n",
       "      <td>0.201360</td>\n",
       "      <td>0.327050</td>\n",
       "      <td>0.223887</td>\n",
       "      <td>0.487693</td>\n",
       "      <td>0.201360</td>\n",
       "      <td>0.370486</td>\n",
       "      <td>0.300793</td>\n",
       "      <td>0.479052</td>\n",
       "      <td>0.483651</td>\n",
       "      <td>0.286180</td>\n",
       "      <td>0.314373</td>\n",
       "      <td>0.201360</td>\n",
       "      <td>0.379839</td>\n",
       "      <td>0.338926</td>\n",
       "      <td>0.454716</td>\n",
       "      <td>0.350081</td>\n",
       "      <td>0.492240</td>\n",
       "      <td>0.497949</td>\n",
       "      <td>0.475668</td>\n",
       "      <td>0.500626</td>\n",
       "      <td>0.401057</td>\n",
       "      <td>0.327050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>56.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.600000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>71.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.600000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.600000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>71.000000</td>\n",
       "      <td>84.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Outcome  Survival.Months         Age  Num.Primaries           N           M   Radiation  \\\n",
       "count  190.000000       190.000000  190.000000     190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.789474        22.186842   70.173684       0.226316    1.147368    0.042105    0.331579   \n",
       "std      0.408760        12.420140    6.146909       0.419551    0.860275    0.201360    0.472024   \n",
       "min      0.000000         9.000000   56.000000       0.000000    0.000000    0.000000    0.000000   \n",
       "25%      1.000000        11.000000   67.000000       0.000000    0.000000    0.000000    0.000000   \n",
       "50%      1.000000        16.000000   71.000000       0.000000    1.000000    0.000000    0.000000   \n",
       "75%      1.000000        34.000000   74.000000       0.000000    2.000000    0.000000    1.000000   \n",
       "max      1.000000        71.000000   84.000000       1.000000    3.000000    1.000000    1.000000   \n",
       "\n",
       "       Tumor.Size  Num.Mutated.Genes  Num.Mutations        AKT1    ALK_Col1    ALK_Col2  \\\n",
       "count  190.000000         190.000000     190.000000  190.000000  190.000000  190.000000   \n",
       "mean     4.545789           2.684211       3.084211    0.010526    0.010526    0.010526   \n",
       "std      2.186359           1.460327       1.697575    0.102326    0.102326    0.102326   \n",
       "min      1.000000           0.000000       0.000000    0.000000    0.000000    0.000000   \n",
       "25%      3.600000           2.000000       2.000000    0.000000    0.000000    0.000000   \n",
       "50%      4.600000           3.000000       3.000000    0.000000    0.000000    0.000000   \n",
       "75%      4.600000           3.000000       4.000000    0.000000    0.000000    0.000000   \n",
       "max     10.000000           8.000000       8.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "              APC    ATM_Col1    ATM_Col2        BRAF       CCND2      CDKN2A      CTNNB1  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.100000    0.021053    0.005263    0.005263    0.010526    0.236842    0.021053   \n",
       "std      0.300793    0.143939    0.072548    0.072548    0.102326    0.426268    0.143939   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "           DNMT3A        EGFR       ERBB3       ERBB4        ESR1       FBXW7       FGFR1  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.015789    0.031579    0.005263    0.010526    0.005263    0.042105    0.021053   \n",
       "std      0.124990    0.175338    0.072548    0.102326    0.072548    0.201360    0.143939   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "            FGFR3        FLT4       FOXL2        GNAS       HNF1A   KRAS_Col1   KRAS_Col2  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.010526    0.010526    0.010526    0.036842    0.005263    0.289474    0.005263   \n",
       "std      0.102326    0.102326    0.102326    0.188872    0.072548    0.454716    0.072548   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    1.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "           MAP2K2         MET    MLH_Col2        MSH2        MSH6     NF_Col1     NF_Col2  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.005263    0.047368    0.005263    0.157895    0.036842    0.026316    0.052632   \n",
       "std      0.072548    0.212987    0.072548    0.365606    0.188872    0.160496    0.223887   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "          NF_Col3     NF_Col5      NOTCH1       NTRK1      PDGFRB      PIK3CA      PIK3CB  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.036842    0.010526    0.021053    0.036842    0.026316    0.036842    0.057895   \n",
       "std      0.188872    0.102326    0.143939    0.188872    0.160496    0.188872    0.234161   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "        POLD_Col2       PTCH1        PTEN         RB1     SMARCA4     SMARCB1         SMO  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.026316    0.021053    0.036842    0.021053    0.005263    0.047368    0.042105   \n",
       "std      0.160496    0.143939    0.188872    0.143939    0.072548    0.212987    0.201360   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "            STK11        TERT   TP53_Col1   TP53_Col2        TSC2     Stage_2     Stage_3  \\\n",
       "count  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000  190.000000   \n",
       "mean     0.121053    0.052632    0.615789    0.042105    0.163158    0.100000    0.352632   \n",
       "std      0.327050    0.223887    0.487693    0.201360    0.370486    0.300793    0.479052   \n",
       "min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "50%      0.000000    0.000000    1.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "75%      0.000000    0.000000    1.000000    0.000000    0.000000    0.000000    1.000000   \n",
       "max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   \n",
       "\n",
       "          Stage_4  Primary.Site_Left Lower Lobe  Primary.Site_Left Upper Lobe  Primary.Site_Other  \\\n",
       "count  190.000000                    190.000000                    190.000000          190.000000   \n",
       "mean     0.368421                      0.089474                      0.110526            0.042105   \n",
       "std      0.483651                      0.286180                      0.314373            0.201360   \n",
       "min      0.000000                      0.000000                      0.000000            0.000000   \n",
       "25%      0.000000                      0.000000                      0.000000            0.000000   \n",
       "50%      0.000000                      0.000000                      0.000000            0.000000   \n",
       "75%      1.000000                      0.000000                      0.000000            0.000000   \n",
       "max      1.000000                      1.000000                      1.000000            1.000000   \n",
       "\n",
       "       Primary.Site_Right Hilar  Primary.Site_Right Lower Lobe  Primary.Site_Right Upper Lobe  \\\n",
       "count                190.000000                     190.000000                     190.000000   \n",
       "mean                   0.173684                       0.131579                       0.289474   \n",
       "std                    0.379839                       0.338926                       0.454716   \n",
       "min                    0.000000                       0.000000                       0.000000   \n",
       "25%                    0.000000                       0.000000                       0.000000   \n",
       "50%                    0.000000                       0.000000                       0.000000   \n",
       "75%                    0.000000                       0.000000                       1.000000   \n",
       "max                    1.000000                       1.000000                       1.000000   \n",
       "\n",
       "       Histology_Large-cell carcinoma  Histology_Squamous cell carcinoma     Grade_3     Grade_4  \\\n",
       "count                      190.000000                         190.000000  190.000000  190.000000   \n",
       "mean                         0.142105                           0.405263    0.442105    0.342105   \n",
       "std                          0.350081                           0.492240    0.497949    0.475668   \n",
       "min                          0.000000                           0.000000    0.000000    0.000000   \n",
       "25%                          0.000000                           0.000000    0.000000    0.000000   \n",
       "50%                          0.000000                           0.000000    0.000000    0.000000   \n",
       "75%                          0.000000                           1.000000    1.000000    1.000000   \n",
       "max                          1.000000                           1.000000    1.000000    1.000000   \n",
       "\n",
       "              T_2         T_3         T_4  \n",
       "count  190.000000  190.000000  190.000000  \n",
       "mean     0.526316    0.200000    0.121053  \n",
       "std      0.500626    0.401057    0.327050  \n",
       "min      0.000000    0.000000    0.000000  \n",
       "25%      0.000000    0.000000    0.000000  \n",
       "50%      1.000000    0.000000    0.000000  \n",
       "75%      1.000000    0.000000    0.000000  \n",
       "max      1.000000    1.000000    1.000000  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_imputed.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
